cpu/atmega2560/periph/gpio.c

```diff
@@ -86,51 +86,36 @@ static inline uint16_t _pin_addr(gpio_t pin)
     return (_port_addr(pin) - 0x02);
 }
 
-int gpio_init(gpio_t pin, gpio_dir_t dir, gpio_pp_t pullup)
+int gpio_init(gpio_t pin, gpio_mode_t mode)
 {
-    int res;
-
-    if (dir == GPIO_DIR_OUT) {
+    if (mode == GPIO_OUT) {
         _SFR_MEM8(_ddr_addr(pin)) |= (1 << _pin_num(pin));
-        res = bit_is_set(_SFR_MEM8(_ddr_addr(pin)), _pin_num(pin));
     }
-    else {
+    else if (mode == GPIO_IN) {
         _SFR_MEM8(_ddr_addr(pin)) &= ~(1 << _pin_num(pin));
-        res = bit_is_clear(_SFR_MEM8(_ddr_addr(pin)), _pin_num(pin));
-
-        if (res == 0) {
-            return -1;
-        }
-
-        switch (pullup) {
-            case GPIO_NOPULL:
-                _SFR_MEM8(_port_addr(pin)) &= ~(1 << _pin_num(pin));
-                res = bit_is_clear(_SFR_MEM8(_port_addr(pin)), _pin_num(pin));
-                break;
-            case GPIO_PULLUP:
-                _SFR_MEM8(_port_addr(pin)) |= (1 << _pin_num(pin));
-                res = bit_is_set(_SFR_MEM8(_port_addr(pin)), _pin_num(pin));
-                break;
-            case GPIO_PULLDOWN:
-                /* Not supported by atmega2560 */
-                return -1;
-        }
+        _SFR_MEM8(_port_addr(pin)) &= ~(1 << _pin_num(pin));
     }
-
-    return (res == 0) ? -1 : 0;
+    else if (mode == GPIO_IN_PU) {
+        _SFR_MEM8(_port_addr(pin)) |= (1 << _pin_num(pin));
+    }
+    else {
+        return -1;
+    }
+    return 0;
 }
 
-int gpio_init_int(gpio_t pin, gpio_pp_t pullup, gpio_flank_t flank,
-                  gpio_cb_t cb, void *arg)
+int gpio_init_int(gpio_t pin, gpio_mode_t mode, gpio_flank_t flank,
+                    gpio_cb_t cb, void *arg)
 {
     uint8_t pin_num = _pin_num(pin);
 
     if ((_port_num(pin) == PORT_D && pin_num > 3)
-         || (_port_num(pin) == PORT_E && pin_num < 4)) {
+         || (_port_num(pin) == PORT_E && pin_num < 4)
+         || (mode >= GPIO_OUT)) {
```

Comment:
I don't like this greater than comparison with an enum since it will silently break if the order of the enum values in the global periph/gpio.h file changes. Is there a big penalty on code size if we change this into the below?

```c
switch(mode) {
  case GPIO_IN:
  case GPIO_IN_PU:
  /* (case GPIO_IN_PD:) since atmega2560 does not seem to support pull-down */
  break;
  default:
  return -1;
```